frame/elections-phragmen/src/lib.rs

```diff
@@ -289,210 +434,158 @@ decl_error! {
 		///
-		/// It is the responsibility of the caller to not place all of their balance into the lock
-		/// and keep some for further transactions.
+		/// The dispatch origin of this call must be signed.
+		///
+		/// ### Warning
+		///
+		/// It is the responsibility of the caller to **NOT** place all of their balance into the
+		/// lock and keep some for further transactions.
 		///
 		/// # <weight>
-		/// Base weight: 47.93 µs
-		/// State reads:
-		/// 	- Candidates.len() + Members.len() + RunnersUp.len()
-		/// 	- Voting (is_voter)
-		/// 	- Lock
-		/// 	- [AccountBalance(who) (unreserve + total_balance)]
-		/// State writes:
-		/// 	- Voting
-		/// 	- Lock
-		/// 	- [AccountBalance(who) (unreserve -- only when creating a new voter)]
+		/// we consider the common case of placing more votes. In other two case, we refund.
 		/// # </weight>
-		#[weight = T::WeightInfo::vote(votes.len() as u32)]
+		#[weight = T::WeightInfo::vote_more(votes.len() as u32)]
 		fn vote(
 			origin,
 			votes: Vec<T::AccountId>,
 			#[compact] value: BalanceOf<T>,
-		) {
+		) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
 
+			// votes should not be empty and more than `MAXIMUM_VOTE` in any case.
 			ensure!(votes.len() <= MAXIMUM_VOTE, Error::<T>::MaximumVotesExceeded);
 			ensure!(!votes.is_empty(), Error::<T>::NoVotes);
 
 			let candidates_count = <Candidates<T>>::decode_len().unwrap_or(0);
 			let members_count = <Members<T>>::decode_len().unwrap_or(0);
 			let runners_up_count = <RunnersUp<T>>::decode_len().unwrap_or(0);
 
+			// can never submit a vote of there are no members, and cannot submit more votes than
+			// all potential vote targets.
 			// addition is valid: candidates, members and runners-up will never overlap.
 			let allowed_votes = candidates_count + members_count + runners_up_count;
-
 			ensure!(!allowed_votes.is_zero(), Error::<T>::UnableToVote);
 			ensure!(votes.len() <= allowed_votes, Error::<T>::TooManyVotes);
 
 			ensure!(value > T::Currency::minimum_balance(), Error::<T>::LowBalance);
 
-			// first time voter. Reserve bond.
-			if !Self::is_voter(&who) {
-				T::Currency::reserve(&who, T::VotingBond::get())
-					.map_err(|_| Error::<T>::UnableToPayBond)?;
-			}
+			// Reserve bond.
+			let new_deposit = Self::deposit_of(votes.len());
+			let Voter { deposit: old_deposit, .. } = <Voting<T>>::get(&who);
+			let maybe_refund = match new_deposit.cmp(&old_deposit) {
+				Ordering::Greater => {
+					// Must reserve a bit more.
+					let to_reserve = new_deposit - old_deposit;
+					T::Currency::reserve(&who, to_reserve)
+						.map_err(|_| Error::<T>::UnableToPayBond)?;
+					None
+				},
+				Ordering::Equal => {
+					Some(T::WeightInfo::vote_equal(votes.len() as u32))
+				},
+				Ordering::Less => {
+					// Must unreserve a bit.
+					let to_unreserve = old_deposit - new_deposit;
+					let _remainder = T::Currency::unreserve(&who, to_unreserve);
+					debug_assert!(_remainder.is_zero());
```

Comment:
is this needed? I don't see a benefit...

Reply:
I have a tendency to sprinkle these invariants into the code as debug_assertions and they have saved my arse in the past, so preferably keeping them.